x-pack/metricbeat/module/openai/usage/usage.go

```diff
@@ -0,0 +1,392 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package usage
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"path"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"github.com/elastic/beats/v7/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+	"github.com/elastic/elastic-agent-libs/paths"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host is defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("openai", "usage", New)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	logger     *logp.Logger
+	config     Config
+	report     mb.ReporterV2
+	stateStore *stateStore
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The openai usage metricset is beta.")
+
+	config := defaultConfig()
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+
+	st, err := newStateStore(paths.Resolve(paths.Data, path.Join(base.Module().Name(), base.Name())))
+	if err != nil {
+		return nil, fmt.Errorf("creating state store: %w", err)
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		logger:        logp.NewLogger("openai.usage"),
+		config:        config,
+		stateStore:    st,
+	}, nil
+}
+
+// Fetch method implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+//
+//  1. Creates a rate-limited HTTP client with configured timeout and burst settings
+//  2. Sets up the time range for data collection:
+//     i.	End date is current UTC time for realtime collection
+//     ii.	End date is previous day for non-realtime collection
+//     iii. Start date is calculated based on configured lookback days
+//  3. Fetches usage data for each day in the date range
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+	httpClient := newClient(
+		context.TODO(),
+		m.logger,
+		rate.NewLimiter(
+			rate.Every(time.Duration(*m.config.RateLimit.Limit)*time.Second),
+			*m.config.RateLimit.Burst,
+		),
+		m.config.Timeout,
+	)
```

Comment:
any special reason to create the client on every Fetch? Wouldn't be better to create it on `New` and reuse it when fetch is called?
Also, let's say for any some reason Fetch takes to long to respond and it's called a second time, while the 1st hasn't returned yet, with the current approach, each fetch would have its own rate limit, what is effectively doubling the rate intended rate limit

Reply:
Yes, right. My bad. Moved it to `New`